File: src/NonDecreasingArray.java
public class NonDecreasingArray {
    public boolean checkPossibility(int[] array) {
        boolean handledDecreasing = false;
        for (int index = 0 ; index < array.length - 1 ; index++) {
            if (array[index] > array[index + 1]) {
                if (handledDecreasing) return false;
                handledDecreasing = true;
                if ((index - 1 >= 0 && array[index - 1] > array[index + 1]) && (index + 2 < array.length && array[index] > array[index + 2])) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: python/non_decreasing_array.py
from typing import List


class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        handled_decreasing = False
        for index in range(len(nums) - 1):
            if nums[index] > nums[index + 1]:
                if handled_decreasing: return False
                handled_decreasing = True
                if index - 1 >= 0 and nums[index - 1] > nums[index + 1] and index + 2 < len(nums) and nums[index] > nums[index + 2]:
                    return False
        return True
